contracts/oracle/CollateralizationOracle.sol

```diff
@@ -0,0 +1,289 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.4;
+
+import "./IOracle.sol";
+import "./ICollateralizationOracle.sol";
+import "../refs/CoreRef.sol";
+import "../pcv/IPCVDepositV2.sol";
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+
+interface IPausable {
+    function paused() external view returns (bool);
+}
+
+/// @title Fei Protocol's Collateralization Oracle
+/// @author eswak
+/// @notice Reads a list of PCVDeposit that report their amount of collateral
+///         and the amount of protocol-owned FEI they manage, to deduce the
+///         protocol-wide collateralization ratio.
+contract CollateralizationOracle is ICollateralizationOracle, CoreRef {
+    using Decimal for Decimal.D256;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    // ----------- Events -----------
+
+    event DepositAdd(address from, address indexed deposit, address indexed token);
+    event DepositRemove(address from, address indexed deposit);
+    event OracleUpdate(address from, address indexed token, address indexed oldOracle, address indexed newOracle);
+
+    // ----------- Properties -----------
+
+    /// @notice List of PCVDeposits to exclude from calculations
+    mapping(address => bool) public excludedDeposits;
+
+    /// @notice Map of oracles to use to get USD values of assets held in
+    ///         PCV deposits. This map is used to get the oracle address from
+    ///         and ERC20 address.
+    mapping(address => address) public tokenToOracle;
+    /// @notice Map from token address to an array of deposit addresses. It is
+    //          used to iterate on all deposits while making oracle requests
+    //          only once.
+    mapping(address => EnumerableSet.AddressSet) private tokenToDeposits;
+    /// @notice Map from deposit address to token address. It is used like an
+    ///         indexed version of the pcvDeposits array, to check existence
+    ///         of a PCVdeposit in the current config.
+    mapping(address => address) public depositToToken;
+    /// @notice Array of all tokens held in the PCV. Used for iteration on tokens
+    ///         and oracles.
+    EnumerableSet.AddressSet private tokensInPcv;
+
+    // ----------- Constructor -----------
+
+    /// @notice CollateralizationOracle constructor
+    /// @param _core Fei Core for reference
+    constructor(
+        address _core
+    ) CoreRef(_core) {}
+
+    // ----------- Convenience getters -----------
+
+    /// @notice returns true if a token is held in the pcv
+    function isTokenInPcv(address token) external view returns(bool) {
+        return tokensInPcv.contains(token);
```

Comment:
Aren't we better off looking up a token in the tokenToOracle mapping and ensuring that the resulting address is not address 0 instead of enumerating over a list of addresses?